File: packages/backend/sdk/tests/webhooks.test.ts
import { createHmac } from 'node:crypto';
import { describe, expect, it } from 'vitest';
import { verifyWebhookSignature } from '../src/webhooks.js';

function computeSignature(
  payload: unknown,
  timestamp: string,
  secret: string,
): string {
  function sortKeys(value: unknown): unknown {
    if (Array.isArray(value)) {
      return value.map(sortKeys);
    }
    if (value !== null && typeof value === 'object') {
      return Object.keys(value as object)
        .sort()
        .reduce(
          (acc, key) => {
            acc[key] = sortKeys((value as Record<string, unknown>)[key]);
            return acc;
          },
          {} as Record<string, unknown>,
        );
    }
    return value;
  }

  const canonicalJson = JSON.stringify(sortKeys(payload));
  const signedString = `${timestamp}.${canonicalJson}`;
  return createHmac('sha256', secret).update(signedString).digest('hex');
}

describe('verifyWebhookSignature', () => {
  const secret = 'webhook-test-secret-123';
  const timestamp = () => Math.floor(Date.now() / 1000).toString();

  it('returns true for a valid signature', () => {
    const payload = { id: 'pay-1', status: 'completed' };
    const ts = timestamp();
    const signature = computeSignature(payload, ts, secret);

    expect(verifyWebhookSignature(payload, signature, ts, secret)).toBe(true);
  });

  it('returns true when payload keys are in a different order', () => {
    const payload = { status: 'completed', id: 'pay-1' };
    const ts = timestamp();
    const signature = computeSignature({ id: 'pay-1', status: 'completed' }, ts, secret);

    expect(verifyWebhookSignature(payload, signature, ts, secret)).toBe(true);
  });

  it('returns false for a tampered payload', () => {
    const payload = { id: 'pay-1', status: 'completed' };
    const ts = timestamp();
    const signature = computeSignature(payload, ts, secret);

    expect(
      verifyWebhookSignature({ id: 'pay-1', status: 'failed' }, signature, ts, secret),
    ).toBe(false);
  });

  it('returns false for a wrong secret', () => {
    const payload = { id: 'pay-1', status: 'completed' };
    const ts = timestamp();
    const signature = computeSignature(payload, ts, 'wrong-secret');

    expect(verifyWebhookSignature(payload, signature, ts, secret)).toBe(false);
  });

  it('returns false for an expired timestamp', () => {
    const payload = { id: 'pay-1', status: 'completed' };
    const ts = (Math.floor(Date.now() / 1000) - 11).toString();
    const signature = computeSignature(payload, ts, secret);

    expect(verifyWebhookSignature(payload, signature, ts, secret)).toBe(false);
  });

  it('returns false for an invalid timestamp', () => {
    const payload = { id: 'pay-1', status: 'completed' };
    const signature = computeSignature(payload, timestamp(), secret);

    expect(verifyWebhookSignature(payload, signature, 'not-a-timestamp', secret)).toBe(false);
  });

  it('returns false for mismatched length signatures', () => {
    const payload = { id: 'pay-1' };
    expect(verifyWebhookSignature(payload, 'abc', timestamp(), secret)).toBe(false);
  });

  it('returns false for an empty signature', () => {
    const payload = { id: 'pay-1' };
    expect(verifyWebhookSignature(payload, '', timestamp(), secret)).toBe(false);
  });

  it('handles nested objects with sorted keys', () => {
    const payload = {
      event: 'payment.completed',
      data: { z: 1, a: { y: 2, b: 3 } },
    };
    const ts = timestamp();
    const signature = computeSignature(payload, ts, secret);

    expect(verifyWebhookSignature(payload, signature, ts, secret)).toBe(true);
  });
});

File: packages/backend/sdk-go/fluxis/webhooks_test.go
package fluxis

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"sort"
	"strconv"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

func computeWebhookSignature(payload any, timestamp, secret string) string {
	sorted := sortKeys(payload)
	canonical, _ := json.Marshal(sorted)
	signedString := timestamp + "." + string(canonical)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(signedString))
	return hex.EncodeToString(mac.Sum(nil))
}

func currentTimestamp() string {
	return strconv.FormatInt(time.Now().Unix(), 10)
}

func TestVerifyWebhookSignature(t *testing.T) {
	secret := "webhook-test-secret-123"

	t.Run("valid signature", func(t *testing.T) {
		payload := map[string]any{"id": "pay-1", "status": "completed"}
		ts := currentTimestamp()
		signature := computeWebhookSignature(payload, ts, secret)
		assert.True(t, VerifyWebhookSignature(payload, signature, ts, secret))
	})

	t.Run("different key order", func(t *testing.T) {
		payload := map[string]any{"status": "completed", "id": "pay-1"}
		ts := currentTimestamp()
		signature := computeWebhookSignature(map[string]any{"id": "pay-1", "status": "completed"}, ts, secret)
		assert.True(t, VerifyWebhookSignature(payload, signature, ts, secret))
	})

	t.Run("tampered payload", func(t *testing.T) {
		payload := map[string]any{"id": "pay-1", "status": "completed"}
		ts := currentTimestamp()
		signature := computeWebhookSignature(payload, ts, secret)
		assert.False(t, VerifyWebhookSignature(map[string]any{"id": "pay-1", "status": "failed"}, signature, ts, secret))
	})

	t.Run("wrong secret", func(t *testing.T) {
		payload := map[string]any{"id": "pay-1", "status": "completed"}
		ts := currentTimestamp()
		signature := computeWebhookSignature(payload, ts, "wrong-secret")
		assert.False(t, VerifyWebhookSignature(payload, signature, ts, secret))
	})

	t.Run("expired timestamp", func(t *testing.T) {
		payload := map[string]any{"id": "pay-1", "status": "completed"}
		ts := strconv.FormatInt(time.Now().Unix()-11, 10)
		signature := computeWebhookSignature(payload, ts, secret)
		assert.False(t, VerifyWebhookSignature(payload, signature, ts, secret))
	})

	t.Run("invalid timestamp", func(t *testing.T) {
		payload := map[string]any{"id": "pay-1", "status": "completed"}
		signature := computeWebhookSignature(payload, currentTimestamp(), secret)
		assert.False(t, VerifyWebhookSignature(payload, signature, "not-a-timestamp", secret))
	})

	t.Run("mismatched signature length", func(t *testing.T) {
		payload := map[string]any{"id": "pay-1"}
		assert.False(t, VerifyWebhookSignature(payload, "abc", currentTimestamp(), secret))
	})

	t.Run("empty signature", func(t *testing.T) {
		payload := map[string]any{"id": "pay-1"}
		assert.False(t, VerifyWebhookSignature(payload, "", currentTimestamp(), secret))
	})

	t.Run("nested sorted keys", func(t *testing.T) {
		payload := map[string]any{
			"event": "payment.completed",
			"data": map[string]any{
				"z": 1,
				"a": map[string]any{"y": 2, "b": 3},
			},
		}
		ts := currentTimestamp()
		signature := computeWebhookSignature(payload, ts, secret)
		assert.True(t, VerifyWebhookSignature(payload, signature, ts, secret))
	})
}

func TestSortKeys(t *testing.T) {
	input := map[string]any{"b": 2, "a": 1}
	sorted := sortKeys(input).(map[string]any)
	keys := make([]string, 0, len(sorted))
	for key := range sorted {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	assert.Equal(t, []string{"a", "b"}, keys)
}
